Feature/5 add user join

jieun5119 opened the pull request:
## 👩‍💻 요구 사항과 구현 내용 
1. 이메일 중복 검사
2. 닉네임 중복 검사
3. 핸드폰 번호 중복 검사
4. 비밀번호 확인 
## ✅ 피드백 반영사항  
- 이메일 유효성 검사 메시지 오타 수정
- fromEntity 메서드 이름 변경
- UserService 내 예외 처리 추가
- 비즈니스 로직 Service로 이동
- Controller 엔드 포인트 보관 (코드 내 삭제)
- User 타입 명시적 선언
- 중복 체크 로직을 findBy에서 existsBy로 수정 / 메소드명 / 접근제어자 수정

## ✅ PR 포인트 & 궁금한 점 
- 이메일, 닉네임, 핸드폰 번호 중복 검사를 간단하게 구현했는데 혹시 피드백 주실 부분 있으시면 피드백 부탁 드립니다.
- 아마 부족한 부분이 많을 텐데 편하게 피드백 해주시면 감사하겠습니다.

jhk01007 committed:
docs: Create PULL_REQUEST_TEMPLATE.md

PR 템플릿 업데이트

jhk01007 committed:
docs: PULL_REQUEST_TEMPLATE.md 삭제

디렉토리 경로 잘못잡아서 삭제합니다.

jhk01007 committed:
docs: PULL_REQUEST_TEMPLATE.md 생성

PR 템플릿 생성

7zrv committed:
feat: 쓰레기통 등록 API 구현

- 쓰레기통 등록을 위한 컨트롤러 엔드포인트 구현
- 입력 검증 및 데이터 전송을 위한 record 타입의 RegisterTrashcanRequestDto 추가
- 요청DTO에 위도, 경도 및 주소 필드에 대한 검증 어노테이션 추가
- 검증 실패에 대한 예외 처리 포함
- 쓰레기통 등록 처리를 위한 TrashcanService 메서드 생성
- 쓰레기통 엔티티 DB 연결을 위한 Repository 생성

7zrv committed:
feat: id를 통한 쓰레기통 조회 API 구현

- 쓰레기통 조회를 위한 컨트롤러 엔드포인트 구현
- 검색 실패에 대한 예외 처리 포함
- 쓰레기통 등록 처리를 위한 TrashcanService 메서드 생성

7zrv committed:
Fix: Enum, Validation 어노테이션 수정

리뷰 해주신 부분들을 수정했습니다.
- TrasgcanStatus의 소문자로 표기된 값 대문자로 수정
- RegisterTrashcanRequestDto의 validation 어노테이션 변경

7zrv committed:
Fix: 예외처리코드 enum 수정

- NOT_FOUND 에러 코드를 NOT_EXIST로 수정

jieun5119 committed:
feat: 회원가입 기능 구현 \n\n이메일,닉네임,전화번호 중복확인\n비밀번호 확인

# Conflicts:
#	src/main/java/com/sscanner/team/global/exception/ExceptionCode.java

jieun5119 committed:
Feat: 회원가입 기능 구현
- 이메일, 닉네임, 휴대폰번호 중복 체크
- 비밀번호 확인

jieun5119 committed:
Feat: 회원가입 기능 구현

7zrv commented on src/main/java/com/sscanner/team/user/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+    @ResponseBody
+    public ApiResponse<String> registerUSer(){
+        return ApiResponse.ok(200, "회원가입 API. POST 요청으로 가입 정보 전송.", "회원가입 페이지 이동");    }
+
```
이 부분의 엔드포인트는 따로 보관을 하시다가 나중에 추가를 해보는건 어떨까요?
만약 프론트 서버를 만들게 된다면 불필요한 엔드포인트가 될 것 같고
서버사이드 렌더링을 하게된다면 view controller가 생길것 같은데
그때 view controller에 추가해주면 좋을것 같습니다!

jieun5119 replied:
넵 무슨 말씀인지 이해했습니다 ! 수정하도록 하겠습니다

7zrv commented on src/main/java/com/sscanner/team/user/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+
+        // 회원가입
+        UserJoinResponseDto responseDTO = userService.join(requestDTO);
+
```
제 생각에는 중복 체크 로직들을 UserService의 join 메서드에 넣어주면 더 깔끔해질거 같아요
컨트롤러는 요청을 받고 응답을 내준다는 역할에만 충실한게 좋다고 생각해요!

jieun5119 replied:
넵 감사합니다 컨트롤러의 역할이 더 명확해지도록 수정해보겠습니다 !

7zrv commented on src/main/java/com/sscanner/team/user/requestDto/UserJoinRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+
+        @NotBlank(message = "에미일이 비어있습니다.")
+        String email,
+
```
메세지에 오타가 있는것 같습니다!

7zrv commented on src/main/java/com/sscanner/team/user/responseDto/UserJoinResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+        String phone
+){
+
+    public static UserJoinResponseDto fromEntity(User user) {
```
제 생각에 from이라고 이름지어도 충분히 직관적일것 같습니다!

7zrv commented on src/main/java/com/sscanner/team/user/service/UserService.java:
```diff
@@ -0,0 +1,49 @@
+    public boolean checkPhone(String phone) {
+        return userRepository.existsByPhone(phone);
+    }
+
```
return userRepository.existsByPhone(phone).orElseThrow(() -> new ~~~~~~);
이런식으로 메서드 자체에서 예외처리를 해주면 더 좋은 코드가 될 것 같아요!

jieun5119 replied:
넵! 메서드 내에 예외처리 추가하도록 하겠습니다 !

7zrv commented:
main -> develop으로 병합할 곳 변경 해야할 것 같습니다!

jieun5119 committed:
refactor: 이메일 유효성 검사 메시지 수정

jieun5119 committed:
refactor: 메서드 이름 변경

fromEntity 메서드를 from으로 이름 변경하여 가독성 개선

jieun5119 committed:
refactor: 회원가입 기능 리팩토링

- 중복 체크 로직을 UserService의 join 메서드로 이동\n- UserRepository 수정\n- UserService에 직접 예외 처리 추가\n- 비즈니스 로직을 서비스 계층으로 이동하고, 컨트롤러는 요청과 응답에 집중\n- 엔드포인트 보관

7zrv commented on src/main/java/com/sscanner/team/user/service/UserService.java:
```diff
@@ -14,33 +17,50 @@ public class UserService {
+    public void checkEmail(String email) {
+        userRepository.findByEmail(email)
+                .ifPresent(user -> {
+                    throw new DuplicateException(ExceptionCode.DUPLICATED_EMAIL);
```
아마 처음에 exist를 하셨을때 Boolean 타입으로 값이 나와서 find로 수정후 데이터를 받고 ifPresent로 예외처리를 하셨을거 같은데 
제 생각이 맞을까요..?

데이터가 있는지 없는지만 검증하면 되기때문에 저는 처음에 하셨던 exists가 더 좋은 방법 같아요
아래와 같은 방법으로 하면 exists로 검증할 수 잇습니다!
메서드 이름도 중복체크라는 의미를 넣어주면 좋을것 같아요!

private void checkDuplicatedEmail(final String email) {
        if (userRepository.existsByEmail(email)) {
            throw new DuplicateException(DUPLICATED_USER_EMAIL);
        }
    }

추가로 중복처리의 경우 외부로 노출되는 메서드가 아니기 때문에 private으로 접근제어자를 변경해주면 
더 안정적인 클래스가 될 것 같아요!

7zrv commented on src/main/java/com/sscanner/team/user/service/UserService.java:
```diff
@@ -14,33 +17,50 @@ public class UserService {
+        checkPhone(req.phone());
+        confirmPassword(req.password(), req.passwordCheck());
+
         var userEntity = req.toEntity(passwordEncoder.encode(req.password()));
```
var타입을 이용하신 이유를 알려주실수 있으신가요? 한번도 사용을 안해봐서 궁금합니다!

저는 toEntity 라는건 dto를 어떤 객체로 바꿔주는 메서드로 이해를 하고 있는데요
var보다는 User로 타입을 명확하게 해주면 더 좋을것 같습니다!

jieun5119 committed:
Merge branch 'develop' into feature/5-add-user-join

Leehunil commented on src/main/java/com/sscanner/team/user/service/UserService.java:
```diff
@@ -0,0 +1,69 @@
+        checkPhone(req.phone());
+        confirmPassword(req.password(), req.passwordCheck());
+
+        var userEntity = req.toEntity(passwordEncoder.encode(req.password()));
```
타입 추측이 어렵고 가독성이 떨어진다는 이유로 var 타입 사용을 지양한다는 얘기를 들어적이 있습니다!! 여기서 var 타입을 사용하신 이유가 궁금합니다~~

jieun5119 replied:
여기서 저는 변수 타입이 명확하다고 생각해서 var을 사용했습니다.
저는 평소에 var을 코드가 간결해질 수 있다는 장점 때문 가끔 썼었는데 훈일님이 말씀한 것처럼 가독성과 타입 추측에 문제가 있을 수 있을 것 같네요.  명시적인 타입을 사용하는 방향으로 수정하겠습니다 !

jieun5119 committed:
fix: User import 및 타입 명시적 선언

jieun5119 committed:
fix:ExceptionCode 수정

7zrv commented on src/main/java/com/sscanner/team/user/repository/UserRepository.java:
```diff
@@ -3,8 +3,12 @@
-    boolean existsByPhone(String phone);
+    Optional<User> findByEmail(String email);
+    Optional<User> findByNickname(String nickname);
+    Optional<User> findByPhone(String phone);
```
중복처리의 경우 값이 있냐 없냐를 판단하기만 하면된다고 생각해서 원래 하셨던 exists가 조금더 효율적이지 않을까 생각했습니다!

jieun5119 committed:
fix: 중복 체크 로직을 findBy에서 existsBy로 수정
